src/kernels/MIOpenFractionalMaxPool.cpp

```diff
@@ -0,0 +1,345 @@
+                if(val > m || isnan(val))
+                {
+                    m = val;
+                }
+            }
+        }
+        output[output_tv.get_tensor_view_idx(layout)] = CVT_ACCUM2FLOAT(m);
+    }
+    else
+    {
+        int64_t mi = pool_h * input_tv.size[3] + pool_w;
+        for(int64_t h = pool_h; h < h_end; ++h)
+        {
+            for(int64_t w = pool_w; w < w_end; ++w)
+            {
+                FLOAT_ACCUM val =
+                    CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({n, c, h, w})]);
+                if(val > m || isnan(val))
+                {
+                    m  = val;
+                    mi = h * input_tv.size[3] + w;
+                }
+            }
+        }
+
+        output[output_tv.get_tensor_view_idx(layout)]   = CVT_ACCUM2FLOAT(m);
+        indices[indices_tv.get_tensor_view_idx(layout)] = static_cast<Ti>(mi);
+    }
+}
+
+extern "C" __global__ void FractionalMaxPool2dForward(const D_TYPE* input,
+                                                      D_TYPE* output,
+                                                      I_TYPE* indices,
+                                                      D_TYPE* random_sample,
+                                                      int64_t KH,
+                                                      int64_t KW,
+                                                      tensor_view_t<4> input_tv,
+                                                      tensor_view_t<4> output_tv,
+                                                      tensor_view_t<4> indices_tv,
+                                                      tensor_view_t<3> random_sample_tv)
+{
+    fractionalMaxPool2dForward<D_TYPE, I_TYPE>(input,
+                                               output,
+                                               indices,
+                                               random_sample,
+                                               KH,
+                                               KW,
+                                               input_tv,
+                                               output_tv,
+                                               indices_tv,
+                                               random_sample_tv);
+}
+
+template <typename T, typename Ti>
+__device__ void fractionalMaxPool2dBackward(const Ti* indices,
+                                            const T* output_grad,
+                                            T* input_grad,
+                                            uint64_t numel,
+                                            tensor_view_t<4> indices_tv,
+                                            tensor_view_t<4> output_grad_tv,
+                                            tensor_view_t<4> input_grad_tv)
+{
+    uint64_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if(gid >= numel)
+        return;
+
+    tensor_layout_t<4> layout(output_grad_tv, gid);
```

Comment:
Use below code so you don't need to pass `numel` as a kernel param.
```suggestion
   tensor_layout_t<4> layout(output_grad_tv, gid);
   if (layout.layout[0] >= output_grad_tv.size[0]) return;
```

Reply:
resolved